src/hotspot/os/aix/porting_aix.cpp

```diff
@@ -891,3 +897,267 @@ bool AixMisc::query_stack_bounds_for_current_thread(stackbounds_t* out) {
+      result = ::dlopen(filename, Flags);
+      if (result != nullptr) {
+        g_handletable_used++;
+        g_handletable[i].handle = result;
+        g_handletable[i].inode = libstat.st_ino;
+        g_handletable[i].devid = libstat.st_dev;
+        g_handletable[i].refcount = 1;
+      }
+      else {
+        // error analysis when dlopen fails
+        const char* error_report = ::dlerror();
+        if (error_report == nullptr) {
+          error_report = "dlerror returned no error description";
+        }
+        if (ebuf != nullptr && ebuflen > 0) {
+          snprintf(ebuf, ebuflen - 1, "%s", error_report);
+        }
+      }
+    }
+    pthread_mutex_unlock(&g_handletable_mutex);
+  }
+  return result;
+}
+
+bool os::pd_dll_unload(void* libhandle, char* ebuf, int ebuflen) {
+  int i = 0;
+  bool res = false;
+
+  if (ebuf && ebuflen > 0) {
+    ebuf[0] = '\0';
+    ebuf[ebuflen - 1] = '\0';
+  }
+
+  pthread_mutex_lock(&g_handletable_mutex);
+  // try to find handle in array, which means library was loaded by os::dll_load() call
+  for (i = 0; i < g_handletable_used; i++) {
+    if (g_handletable[i].handle == libhandle) {
+      // handle found, decrease refcount
+      g_handletable[i].refcount--;
+      if (g_handletable[i].refcount > 0) {
+        // if refcount is still >0 then we have to keep library and just return true
+        pthread_mutex_unlock(&g_handletable_mutex);
+        return true;
+      }
+      // refcount == 0, so we have to ::dlclose() the lib
+      // and delete the entry from the array.
+      break;
+    }
+  }
+
+  // If we reach this point either the libhandle was found with refcount == 0, or the libhandle
+  // was not found in the array at all. In both cases we have to ::dlclose the lib and perform
+  // the error handling. In the first case we then also have to delete the entry from the array
+  // while in the second case we simply have to nag.
+  res = (0 == ::dlclose(libhandle));
+  if (!res) {
+    // error analysis when dlopen fails
+    const char* error_report = ::dlerror();
+    if (error_report == nullptr) {
+      error_report = "dlerror returned no error description";
+    }
+    if (ebuf != nullptr && ebuflen > 0) {
+      snprintf(ebuf, ebuflen - 1, "%s", error_report);
+    }
+  #ifdef ASSERT
+    pthread_mutex_unlock(&g_handletable_mutex);
+  #endif
+    assert(false, "os::pd_dll_unload() ::dlclose() failed");
+  }
+
+  if (i < g_handletable_used) {
+    if (res) {
+      // First case: libhandle was found (with refcount == 0) and ::dlclose successful,
+      // so delete entry from array
+      g_handletable_used--;
+      // If the entry was the last one of the array, the previous g_handletable_used--
+      // is sufficient to remove the entry from the array, otherwise we move the last
+      // entry of the array to the place of the entry we want to remove and overwrite it
+      if (i < g_handletable_used) {
+        g_handletable[i] = g_handletable[g_handletable_used];
```

Comment:
To be super careful, I would zero out at least the handle of the moved item like this:
`g_handletable[g_handletable_used].handle = nullptr`

Reply:
Done